### PR TITLE
gomod2nix: init at 1.5.0

### PR DESCRIPTION
**Pull request by adisbladis**
###### Description of changes

This adds gomod2nix, a generator for Nix expressions using Go modules (and the associated build infrastructure).
It has quite some differences to `buildGoModule` which are laid out in https://www.tweag.io/blog/2021-03-04-gomod2nix/.

I don't feel quite ready yet to promote `buildGoApplication` and `mkGoEnv` to first-class top-level attributes so in nixpkgs these will be accessed by `gomod2nix.buildGoApplication` & `gomod2nix.mkGoEnv` respectively.

This is developed external over at https://github.com/tweag/gomod2nix/.

This work has been sponsored by both the NLNet Foundation under the NGI Zero program and by Tweag.

--- 

Due to a mistake this is a new instance of #188035. See that PR for previous discussion.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by adisbladis** [comment]
Repating myself from: https://github.com/NixOS/nixpkgs/pull/188035#issuecomment-1226881009
> I intend to merge this within the next couple of days unless there are any objections.

**Comment by zowoq** [comment]
The update script seems broken, doesn't seem to eval.  I said previously I won't block on this so if you just want to drop it from this PR and deal with it in a follow up I'm fine with that.

Can you share your plans and time frame for promoting for this?

**Comment by adisbladis** [comment]
> The update script seems broken, doesn't seem to eval. I said previously I won't block on this so if you just want to drop it from this PR and deal with it in a follow up I'm fine with that.

I made a fix to make the updater work in the case that `pname` isn't explicitly passed.

> Can you share your plans and time frame for promoting for this?

I have patches ready for all the Go packages I maintain in nixpkgs and will announce the availability of gomod2nix in nixpkgs shortly after at least a few of those have been merged.
I will however _not_ try to go on some crusade to get everyone to adopt it immediately.

Additionally I also have patches for most of the packages that still use vgo2nix ready too, it's just blocked on this PR.

**Comment by adisbladis** [comment]
@zowoq FYI the upstream repo has been moved to nix-community.

**Comment by zowoq** [comment]
I think the case-sensitive problems in https://github.com/NixOS/nixpkgs/issues/129730 may also be a problem for this?
___

> I made a fix to make the updater work in the case that pname isn't explicitly passed.

Sorry, still doesn't seem to work. Can you add an example to this PR? 

> I have patches ready for all the Go packages I maintain in nixpkgs

Not any of the podman packages please.

> will announce the availability of gomod2nix in nixpkgs shortly after at least a few of those have been merged.

Can we wait at least a couple weeks on this? I'd like to do some testing and fix some nits first. I really don't see why this needs to be rushed, at one point I thought the upstream project had been abandoned. IIRC the go no vendor patch was added 18 months ago.

> FYI the upstream repo has been moved to nix-community.

Cool, thanks.

**Comment by adisbladis** [comment]
> Can we wait at least a couple weeks on this? I'd like to do some testing and fix some nits first. I really don't see why this needs to be rushed

Sure that could wait, no problem.

> Not any of the podman packages please.

I was talking about the ones where I'm the sole maintainer,.

> I think the case-sensitive problems in https://github.com/NixOS/nixpkgs/issues/129730 may also be a problem for this?

I don't think so, I could add a test case for it though.

> Sorry, still doesn't seem to work. Can you add an example to this PR?

Sure, I've added repackaged a fairly low impact package in this PR.

File: pkgs/tools/misc/trillian/gomod2nix.toml [hunk 1]
```diff
@@ -0,0 +1,167 @@
+schema = 3
+subPackages = ["cmd/trillian_log_server", "cmd/trillian_log_signer", "cmd/createtree", "cmd/deletetree", "cmd/updatetree"]
```
Comment (zowoq) [inline pkgs/tools/misc/trillian/gomod2nix.toml:2]: Why does this need to be moved?

File: pkgs/tools/misc/trillian/gomod2nix.toml [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+goPackagePath = "github.com/google/trillian"
```
Comment (zowoq) [inline pkgs/tools/misc/trillian/gomod2nix.toml:3]: - Why is this needed? 
- Why can't we infer this from the `src`? 
- Why does it need to be in this file?

**Comment by zowoq** [comment]
oh the whole thing gets stuffed into the toml. src, version and everything. TBH I really don't like this:

```
{ lib
, gomod2nix
, fetchFromGitHub
}:

gomod2nix.buildGoApplication {
  pname = "trillian";
  pwd = ./.;
  meta = {
    homepage = "https://github.com/google/trillian";
    description = "A transparent, highly scalable and cryptographically verifiable data store.";
    license = [ lib.licenses.asl20 ];
    maintainers = [ lib.maintainers.adisbladis ];
  };
}

```

File: pkgs/tools/misc/trillian/gomod2nix.toml (hunk 1 above)
Comment (adisbladis) [reply pkgs/tools/misc/trillian/gomod2nix.toml:2]: It's not moved per se, you can still set the same attribute in the Nix expression.
This is just another way to set the same thing which is a lot easier to control from the code generator.

File: pkgs/tools/misc/trillian/gomod2nix.toml (hunk 2 above)
Comment (adisbladis) [reply pkgs/tools/misc/trillian/gomod2nix.toml:3]: There is no `src` attribute explicitly being set. 
The `goPackagePath` is there so the derivation knows which sources to select for the `src` attribute.

You could separate the `src` if you wanted to and do the whole manual hash update dance, nothing stops you from doing that but that will leave you without an update script.

**Comment by adisbladis** [comment]
> oh the whole thing gets stuffed into the toml. src, version and everything. TBH I really don't like this:

You _could_ rewrite it in a more verbose way if you wanted to. Look at how https://github.com/nix-community/gomod2nix/blob/master/generic.nix is constructed for example.

There are very good reasons to do it this way though. In `buildGoModule` `src` and `vendorSha256` requires you to update multiple hashes which is absolutely terrible UX.
I didn't want to repeat this terrible UX and made the code generated file the _only_ thing you need to update as a single atomic unit. 

Otherwise we'd have a similarly bad user experience once again for very little reason.

**Comment by zowoq** [comment]
To get the package version I either need to query the package via nix or grep the toml?

**Comment by adisbladis** [comment]
> To get the package version I either need to query the package via nix or grep the toml?

In this instance, yes.

**Comment by zowoq** [comment]
Sorry but I think burying the version in the toml is absolutely horrible.

**Comment by adisbladis** [comment]
> Sorry but I think burying the version in the toml is absolutely horrible.

Ok, I don't see the problem. There is plenty of prior art in nixpkgs already.

**Comment by zowoq** [comment]
> There is plenty of prior art in nixpkgs already.

Package sets or separate packages like this?

> I don't see the problem. 

Searching "github.com/google/trillian" to find the version isn't a problem?  Seems like we're trading one form of bad UX for another?

trillian

**Comment by adisbladis** [comment]
> Package sets or separate packages like this?

Both. Ruby packages comes to mind.

> Searching "github.com/google/trillian" to find the version isn't a problem? Seems like we're trading one form of bad UX for another?

Would you feel better if we also stuck the version attribute in the top-level attributes in the toml file?

**Comment by zowoq** [comment]
> Ruby packages comes to mind.

If you mean `bundlerEnv` with gemset.nix and gemfile.lock that seems very different. Any others? I'm actually curious as I can't think of any. haskell, emacs, vim, node all seem much easier to inspect visually.

> Would you feel better if we also stuck the version attribute in the top-level attributes in the toml file?

Only marginally. Would need to be enforced, I really don't want debates over it and packages flipping back and forth between (buried in toml|top of toml|set in default.nix).

> You could rewrite it in a more verbose way if you wanted to. Look at how https://github.com/nix-community/gomod2nix/blob/master/generic.nix or https://github.com/crypto-org-chain/chain-main/blob/master/default.nix are constructed for example.

I'm kinda worried about the fact that this is possible. At the moment there is (basically) one way of writing a go package, this seems to allow a lot more variation (and more debates). Consistency is good UX also.

**Comment by adisbladis** [comment]
> Any others? I'm actually curious as I can't think of any. haskell, emacs, vim, node all seem much easier to inspect visually.

Carnix is another one (code generator), so is poetry2nix (reads pyproject.toml).

I think it's disingenuous to say that Emacs is easier to inspect visually, the majority of that lives in massive auto generated files. The same goes for Haskell, Node & many more.

My take is that this comment is just bike shedding.

> I'm kinda worried about the fact that this is possible. At the moment there is (basically) one way of writing a go package, this seems to allow a lot more variation (and more debates).

Which is intentional. The needs of a packager packaging up an application in a private repo has somewhat different concerns than someone packaging something that is developed entirely outside of the package tree.

`buildGoModule` lacks this flexibility which (imo) makes it unusable as a tool to integrate into a typical development workflow.

`gomod2nix` was designed with both use cases in mind.

**Comment by SuperSandro2000** [comment]
Did you do some testing with big go.mod files and how they behave like https://github.com/golangci/golangci-lint/blob/master/go.mod and more complex ones like https://github.com/open-policy-agent/gatekeeper/blob/master/go.mod ? Also what if the project uses replace where we can't use ``go install`` with the package name?

File: pkgs/build-support/go/gomod2nix/parser.nix [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+  # Parse lines into a structure
+  parseLines = lines: (foldl'
+    (acc: l:
+      let
+        m = match "([^ )]*) *(.*)" l;
+        directive = elemAt m 0;
+        rest = elemAt m 1;
+
+        # Maintain parser state (inside parens or not)
```
Comment (SuperSandro2000) [inline pkgs/build-support/go/gomod2nix/parser.nix:31]: A wish for the future: some more comments like until here would really help to understand what is being parsed here and increase the bus-factor.

File: pkgs/build-support/go/gomod2nix/symlink/symlink.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	Hash          string `json:"hash"`
+	ReplacedPath  string `json:"replaced,omitempty"`
+}
+
+func main() {
+	sources := make(map[string]string)
+	pkgs := make(map[string]*Package)
+
+	{
+		b, err := ioutil.ReadFile(os.Getenv("sourcesPath"))
```
Comment (SuperSandro2000) [inline pkgs/build-support/go/gomod2nix/symlink/symlink.go:24]:
```suggestion
		b, err := os.ReadFile(os.Getenv("sourcesPath"))
```
https://pkg.go.dev/io/ioutil#example-ReadFile

File: pkgs/build-support/go/gomod2nix/symlink/symlink.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
```
Comment (SuperSandro2000) [inline pkgs/build-support/go/gomod2nix/symlink/symlink.go:6]:
```suggestion
```
ioutil is deprecated https://pkg.go.dev/io/ioutil and using it in new and shiny features is something we probably should not do.

File: pkgs/build-support/go/gomod2nix/symlink/symlink.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+		for _, path := range paths {
+
+			vendorDir := filepath.Join("vendor", filepath.Dir(path))
+			if err := os.MkdirAll(vendorDir, 0755); err != nil {
+				panic(err)
+			}
+
+			if _, err := os.Stat(filepath.Join("vendor", path)); err == nil {
+				files, err := ioutil.ReadDir(src)
```
Comment (SuperSandro2000) [inline pkgs/build-support/go/gomod2nix/symlink/symlink.go:68]: iotuils.ReadDir is also deprecated and replaced by os.ReadDir which returns a slightly different type.

https://pkg.go.dev/io/ioutil#example-ReadFile
```
Deprecated: As of Go 1.16, os.ReadDir is a more efficient and correct choice: it returns a list of fs.DirEntry instead of fs.FileInfo, and it returns partial results in the case of an error midway through reading a directory.
```

File: pkgs/development/tools/gomod2nix/generic.nix [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+}:
+{ stdenv
+, callPackage
+, go
+, lib
+, makeWrapper
+, installShellFiles
+, fetchFromGitHub
+, buildGoApplication ? (callPackage ./builder { }).buildGoApplication
+, mkGoEnv ? (callPackage ./builder { }).buildGoApplication
```
Comment (SuperSandro2000) [inline pkgs/development/tools/gomod2nix/generic.nix:13]: This is not really a great interface. It often causes lots of confusion if a typo is done in the variable names and things just do not work. Could we refactor this to be less error prone?

File: pkgs/development/tools/gomod2nix/src.json [hunk 8]
```diff
@@ -0,0 +1 @@
+{"owner": "tweag", "repo": "gomod2nix", "sha256": "0bgr7h1clfyfl5l25h65nhcw0d0drf77zw0v1p5fjdymiqb0yk8j", "rev": "v1.4.0"}
```
Comment (SuperSandro2000) [inline pkgs/development/tools/gomod2nix/src.json:1]: I am not really a fan of extracting such basic data about a package into a json file. Leaving that inside the package would be my preference. We also have good tooling already to update those information like nix-update. Accessing the attributes from fetchFromGitHub should also be possible otherwise I'd rather passthru more there.

File: pkgs/development/tools/gomod2nix/update.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        dst,
+    ])
+
+
+if __name__ == "__main__":
+
+    pkg_dir = os.path.dirname(os.path.abspath(__file__))
+    nixpkgs_dir = os.path.abspath(os.path.join(pkg_dir, "../../../../"))
+
+    builder_dir = os.path.join(nixpkgs_dir, "pkgs/build-support/go/gomod2nix")
```
Comment (SuperSandro2000) [inline pkgs/development/tools/gomod2nix/update.py:33]: if the nixpkgs root dir is relative then we must already be in this directory, right? Then this could be derived from PWD/CWD and when moving this we don't need to remember to also update the path here.

File: pkgs/development/tools/gomod2nix/update.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env nix-shell
```
Comment (SuperSandro2000) [inline pkgs/development/tools/gomod2nix/update.py:1]: I would have done this script in bash because we are just shelling out to commands but fine for me.

File: pkgs/top-level/all-packages.nix [hunk 11]
```diff
@@ -22220,6 +22220,12 @@ with pkgs;
 
   go2nix = callPackage ../development/tools/go2nix { };
 
+  gomod2nix = let
+    builder = callPackage ../build-support/go/gomod2nix { };
+  in callPackage (import ../development/tools/gomod2nix { inherit lib fetchFromGitHub; }) {
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:22225]: We can't use callPackage here instead of import, can we?

**Comment by SuperSandro2000** [comment]
> oh the whole thing gets stuffed into the toml. src, version and everything. TBH I really don't like this:

same for me. This is not very great from a reviewers perspective who needs to double check that people did not forgot to update version or some hash.

**Comment by FRidh** [comment]
Glad to see a solution without `vendorSha256`!

File: pkgs/build-support/go/gomod2nix/default.nix [hunk 12]
```diff
@@ -0,0 +1,431 @@
...
+      # Install development dependencies from tools.go
+      install = mkInternalPkg "symlink" ./install/install.go;
+
+    };
+
+  fetchGoModule =
+    { hash
+    , goPackagePath
+    , version
+    , go ? pkgs.go
```
Comment (SuperSandro2000) [inline pkgs/build-support/go/gomod2nix/default.nix:55]:
```suggestion
    , go
```
that is always specified in line 100 and having such fallbacks is always a good source of silent errors.

File: pkgs/build-support/go/gomod2nix/default.nix [hunk 13]
```diff
@@ -0,0 +1,431 @@
...
+      goVersion = goMod.go;
+      goAttr = "go_" + (replaceStrings [ "." ] [ "_" ] goVersion);
+    in
+    (
+      if hasAttr goAttr pkgs then pkgs.${goAttr}
+      else trace "go.mod specified Go version ${goVersion} but doesn't exist. Falling back to ${pkgs.go.version}." pkgs.go
+    )
+  ));
+
+  # Strip the rubbish that Go adds to versions, and fall back to a version based on the date if it's a placeholder value
```
Comment (SuperSandro2000) [inline pkgs/build-support/go/gomod2nix/default.nix:144]: ``+incompatible`` is not rubbish but means the go package is not (yet) using go.mod.

**Review by SuperSandro2000 — COMMENTED** [review]
I am sorry that we are doing another review round but I'd rather do this correct the first time than fixing it up multiple times and potentially breaking already packaged packages with it.

File: pkgs/build-support/go/gomod2nix/default.nix [hunk 14]
```diff
@@ -0,0 +1,431 @@
...
+  # Select Go attribute based on version specified in go.mod
+  selectGo = attrs: goMod: attrs.go or (if goMod == null then pkgs.go else
+  (
+    let
+      goVersion = goMod.go;
+      goAttr = "go_" + (replaceStrings [ "." ] [ "_" ] goVersion);
+    in
+    (
+      if hasAttr goAttr pkgs then pkgs.${goAttr}
+      else trace "go.mod specified Go version ${goVersion} but doesn't exist. Falling back to ${pkgs.go.version}." pkgs.go
```
[the remark this answers is not in the document]
Comment (06kellyjac) [reply pkgs/build-support/go/gomod2nix/default.nix:140]: yeah higher compile lower usually works, as long as "Falling back" uses the latest go we have it's probably good

also yeah the version in go.mod is *meant* (sometimes devs don't actually update this) to specify the *lowest* version of go it compiles with
Using a newer one should always be preferable unless something actually breaks

File: pkgs/build-support/go/gomod2nix/default.nix [hunk 15]
```diff
@@ -0,0 +1,431 @@
...
+
+      vendorEnv = mkVendorEnv {
+        inherit go modulesStruct defaultPackage goMod pwd;
+      };
+
+      pname = attrs.pname or (baseNameOf defaultPackage);
+
+    in
+    warnIf (buildFlags != "" || buildFlagsArray != "")
+      "Use the `ldflags` and/or `tags` attributes instead of `buildFlags`/`buildFlagsArray`"
```
[the remark this answers is not in the document]
Comment (06kellyjac) [reply pkgs/build-support/go/gomod2nix/default.nix:248]: this behavior is the same as the existing buildGoModule behavior

I guess if we're hoping to fix some of the other existing pain points we could look through for strings that mention "ldflags" or "tags" which while it could have false positives it should show up a lot less frequently
not sure how the performance would be but I think we can assume there would be a string less than 200 characters or an array of strings less than 30 items in length

File: pkgs/build-support/go/gomod2nix/default.nix (hunk 13 above)
Comment (06kellyjac) [reply pkgs/build-support/go/gomod2nix/default.nix:144]: IIRC it can show up with `+incompatible` even if the project has a `go.mod` as long as there is a mismatch between the module name e.g. `github.com/xyz/xyz` and the tag e.g. `v2.0.3`, to fix it the go.mod needs to be changed to `github.com/xyz/xyz/v2`

File: pkgs/build-support/go/gomod2nix/symlink/symlink.go [hunk 16]
```diff
@@ -0,0 +1,103 @@
+package main
```
Comment (06kellyjac) [inline pkgs/build-support/go/gomod2nix/symlink/symlink.go:1]: I restructured it slightly to make it less indent-hadouken

added a generic function that I have no idea if it actually works :)

moved from `io/ioutil` which go seems to be happy with but needs testing

```go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
)

type Package struct {
	GoPackagePath string `json:"-"`
	Version       string `json:"version"`
	Hash          string `json:"hash"`
	ReplacedPath  string `json:"replaced,omitempty"`
}

type Sources map[string]string

func populateStruct[T any](path string, data T) {
	pathVal := os.Getenv(path)
	if len(path) == 0 {
		panic(fmt.Sprintf("env var '%s' was unset", path))
	}
	path = pathVal

	b, err := os.ReadFile(path)
	if err != nil {
		panic(err)
	}

	err = json.Unmarshal(b, &data)
	if err != nil {
		panic(err)
	}
}

func main() {
	sources := make(Sources)
	pkgs := make(map[string]*Package)

	populateStruct("sourcesPath", sources)

	populateStruct("jsonPath", pkgs)

	keys := make([]string, 0, len(pkgs))
	for key := range pkgs {
		keys = append(keys, key)
	}
	sort.Strings(keys)

	makeSymlinks(keys, sources)
}

func makeSymlinks(keys []string, sources Sources) {
	// Iterate, in reverse order
	for i := len(keys) - 1; i >= 0; i-- {
		key := keys[i]
		src := sources[key]

		paths := []string{key}

		for _, path := range paths {
			vendorDir := filepath.Join("vendor", filepath.Dir(path))
			if err := os.MkdirAll(vendorDir, 0o755); err != nil {
				panic(err)
			}

			vendorPath := filepath.Join("vendor", path)
			if _, err := os.Stat(vendorPath); err == nil {
				populateVendorPath(vendorPath, src)
				continue
			}

			// If the file doesn't already exist, just create a simple symlink
			err := os.Symlink(src, vendorPath)
			if err != nil {
				panic(err)
			}
		}
	}
}

func populateVendorPath(vendorPath string, src string) {
	files, err := os.ReadDir(src)
	if err != nil {
		panic(err)
	}

	for _, f := range files {
		innerSrc := filepath.Join(src, f.Name())
		dst := filepath.Join(vendorPath, f.Name())
		if err := os.Symlink(innerSrc, dst); err != nil {
			// assume it's an existing directory, try to link the directory content instead.
			// TODO should we do this recursively?
			files, err := os.ReadDir(innerSrc)
			if err != nil {
				panic(err)
			}
			for _, f := range files {
				srcFile := filepath.Join(innerSrc, f.Name())
				dstFile := filepath.Join(dst, f.Name())
				if err := os.Symlink(srcFile, dstFile); err != nil {
					fmt.Println("ignoring symlink error", srcFile, dstFile)
				}
			}
		}
	}
}
```

File: pkgs/development/tools/gomod2nix/update.py (hunk 10 above)
Comment (06kellyjac) [reply pkgs/development/tools/gomod2nix/update.py:1]: agreed. there is some JSON that's handled and a tiny bit of JSON creation but nothing that isn't trivial with `jq`

although I'm less inclined to keep it python personally

File: pkgs/build-support/go/gomod2nix/default.nix [hunk 17]
```diff
@@ -0,0 +1,431 @@
...
+  inherit (builtins) substring toJSON hasAttr trace split readFile elemAt;
+  inherit (lib)
+    concatStringsSep replaceStrings removePrefix optionalString pathExists
+    optional concatMapStrings fetchers filterAttrs mapAttrs mapAttrsToList
+    warnIf optionalAttrs platforms
+    ;
+
+  parseGoMod = import ./parser.nix;
+
+  removeExpr = refs: ''remove-references-to ${concatMapStrings (ref: " -t ${ref}") refs}'';
```
Comment (06kellyjac) [inline pkgs/build-support/go/gomod2nix/default.nix:26]:
```suggestion
  removeExpr = refs: "remove-references-to ${concatMapStrings (ref: " -t ${ref}") refs}";
```

the other normal quotes are within `${}` so it should be fine to just use normal quotes

File: pkgs/build-support/go/gomod2nix/default.nix [hunk 18]
```diff
@@ -0,0 +1,431 @@
...
+  internal =
+    let
+      mkInternalPkg = name: src: pkgsBuildBuild.runCommand "gomod2nix-${name}"
+        {
+          inherit (pkgsBuildBuild.go) GOOS GOARCH;
+          nativeBuildInputs = [ pkgsBuildBuild.go ];
+        } ''
+        export HOME=$(mktemp -d)
+        cp ${src} src.go
+        go build -o $out src.go
```
Comment (06kellyjac) [inline pkgs/build-support/go/gomod2nix/default.nix:38]:
```suggestion
        go build -o $out ${src}
```

is the `cp` necessary?

File: pkgs/tools/misc/trillian/gomod2nix.toml (hunk 1 above)
Comment (adisbladis) [reply pkgs/tools/misc/trillian/gomod2nix.toml:2]: For one it serves to reduce the build time closure size.
We don't actually need everything in the root level go.mod and by making the generator aware of what needs to be built we only need to pull in the actually required subset of dependencies.

File: pkgs/tools/misc/trillian/gomod2nix.toml [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+
+[mod]
+  [mod."bitbucket.org/creachadair/shell"]
+    version = "v0.0.7"
+    hash = "sha256-8tCzjb26WacDChPL08BLPiAPSwspr+GYJGiSMHVPdA8="
```
[the remark this answers is not in the document]
Comment (adisbladis) [reply pkgs/tools/misc/trillian/gomod2nix.toml:8]: I don't think it really matters and I don't see anywhere in the TOML library I use where this can be controlled.

File: pkgs/build-support/go/gomod2nix/default.nix (hunk 15 above)
[the remark this answers is not in the document]
Comment (adisbladis) [reply pkgs/build-support/go/gomod2nix/default.nix:248]: Since this is based on the existing `buildGoModule` code I think it's fine.

File: pkgs/build-support/go/gomod2nix/parser.nix (hunk 3 above)
Comment (adisbladis) [reply pkgs/build-support/go/gomod2nix/parser.nix:31]: I've littered this file with as many comments as I think make sense now.

File: pkgs/development/tools/gomod2nix/generic.nix (hunk 7 above)
Comment (adisbladis) [reply pkgs/development/tools/gomod2nix/generic.nix]: I've reworked it and just accepted the duplication and possible divergence in expressions between upstream and nixpkgs instead of trying to maximally share.

The code is much cleaner this way.

File: pkgs/development/tools/gomod2nix/update.py (hunk 10 above)
Comment (adisbladis) [reply pkgs/development/tools/gomod2nix/update.py]: I've ported the update script to bash and it's now using gomod2nix to update itself.

File: pkgs/build-support/go/gomod2nix/default.nix (hunk 14 above)
[the remark this answers is not in the document]
Comment (adisbladis) [reply pkgs/build-support/go/gomod2nix/default.nix]: You're completely right and I've changed my mind about how this functionality works:
It now works as an assertion rather than selecting whatever go.mod specifies.

It will default to the highest compatible version and throw an error if the minimum version specified in go.mod can't be satisfied instead.

File: pkgs/build-support/go/gomod2nix/default.nix (hunk 13 above)
Comment (adisbladis) [reply pkgs/build-support/go/gomod2nix/default.nix]: Changed the comment to convey the true motivation of why it's doing what it's doing rather than just stating what it's doing.

File: pkgs/build-support/go/gomod2nix/symlink/symlink.go (hunk 16 above)
Comment (adisbladis) [reply pkgs/build-support/go/gomod2nix/symlink/symlink.go:1]: Thanks for this, it's much cleaner!

As I want this to be compatible with lower Go versions too I've slightly changed the type signatures not to use generics.

File: pkgs/build-support/go/gomod2nix/default.nix (hunk 18 above)
Comment (adisbladis) [reply pkgs/build-support/go/gomod2nix/default.nix]: No, it's not. Changed.

File: pkgs/build-support/go/gomod2nix/default.nix (hunk 17 above)
Comment (adisbladis) [reply pkgs/build-support/go/gomod2nix/default.nix]: This entire line is unused now anyway so I've removed it.

File: pkgs/development/tools/gomod2nix/update.py (hunk 9 above)
Comment (adisbladis) [reply pkgs/development/tools/gomod2nix/update.py]: The script has been ported to bash.

**Comment by adisbladis** [comment]
Trying this again: If there is nothing new and actionable by Friday (2022-09-09) I will merge this.

**Comment by zowoq** [comment]
> Trying this again: If there is nothing new and actionable by Friday (2022-09-09) I will merge this.

You haven’t responded to a few concerns from Sandro and myself. Between us we’ve done the majority of the go maintenance for the last year so I don’t think our concerns should be dismissed.

> > There is no `src` attribute explicitly being set.

> I am not really a fan of this. If I want to access the source files nix is building things from to investigate and debug packages I often use ``nix build .#name.src`` and I would tell people to do that when debugging. Could we add it at least as an alias back to make things more uniform with other packages?

I don’t think this has been addressed?

> Also we need to make sure that it is clear in the documentation that goPackagePath here has nothing to do with buildGoPackage.

Renaming `goPackagePath` to `goModPath`, `goSrcPath` or something else would eliminate any confusion with `buildGoPackage`.

> what is your plan for documentation?

I don’t mind if this PR is merged without docs but I think having docs is a blocker for `gomod2nix`, etc being promoted.

> > oh the whole thing gets stuffed into the toml. src, version and everything. TBH I really don't like this:

> same for me. This is not very great from a reviewers perspective who needs to double check that people did not forgot to update version or some hash.

This hasn’t been addressed. As you asked for something actionable: I’d like to see the src/hash/version not treated as just another dep and be moved to the top of the toml with the other attributes.

I consider this a blocker for merging this PR.

> Did you do some testing with big go.mod files and how they behave like https://github.com/golangci/golangci-lint/blob/master/go.mod and more complex ones like https://github.com/open-policy-agent/gatekeeper/blob/master/go.mod ? Also what if the project uses replace where we can't use go install with the package name?

Sandro's asked about this twice (in this PR and https://github.com/NixOS/nixpkgs/pull/188035#issuecomment-1227681465) and there doesn't seem to have been a response?

**Comment by 06kellyjac** [comment]
As someone who's also been packaging go modules a bunch I agree with all @zowoq & @SuperSandro2000 's points in that comment

---

I'll try this out with cosign which has a bajillion dependencies

File: pkgs/build-support/go/gomod2nix/install/install.go [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+	"fmt"
+	"go/parser"
+	"go/token"
+	"io"
+	"os"
+	"os/exec"
+	"strconv"
+)
+
+const filename = "tools.go"
```
Comment (06kellyjac) [inline pkgs/build-support/go/gomod2nix/install/install.go:13]: Is it actually enforced that the file is called `tools.go` or just that the build tag is `tools`?

https://marcofranssen.nl/manage-go-tools-via-go-modules

`// +build tools`

I can't really find any docs on this
